x-pack/plugins/apm/public/components/fleet_integration/apm_policy_form/settings/utils.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { isRight } from 'fp-ts/lib/Either';
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import { isEmpty } from 'lodash';
+import { PackagePolicyVars } from '../typings';
+import { SettingDefinition } from './typings';
+
+export const REQUIRED_LABEL = i18n.translate(
+  'xpack.apm.fleet_integration.settings.requiredLabel',
+  { defaultMessage: 'Required' }
+);
+export const OPTIONAL_LABEL = i18n.translate(
+  'xpack.apm.fleet_integration.settings.optionalLabel',
+  { defaultMessage: 'Optional' }
+);
+const REQUIRED_FIELD = i18n.translate(
+  'xpack.apm.fleet_integration.settings.requiredFieldLabel',
+  { defaultMessage: 'Required field' }
+);
+
+export function mergeNewVars(
+  oldVars: PackagePolicyVars,
+  key: string,
+  value?: any
+): PackagePolicyVars {
+  return { ...oldVars, [key]: { ...oldVars[key], value } };
+}
+
+export function isSettingsFormValid(
+  settings: SettingDefinition[],
+  vars: PackagePolicyVars
+) {
+  return settings
+    .filter((field) => field.required || field.validation)
+    .every((field) => {
+      const { value } = vars[field.key];
+      return validateSettingValue(field, value).isValid;
+    });
+}
+
+export function validateSettingValue(setting: SettingDefinition, value?: any) {
+  if (isEmpty(value)) {
+    return {
+      isValid: !setting.required,
+      message: setting.required ? REQUIRED_FIELD : '',
+    };
+  }
+
+  if (setting.validation) {
+    const result = setting.validation.decode(String(value));
+    const message = PathReporter.report(result)[0];
+    const isValid = isRight(result);
```

Comment:
how does `isRight` check if it's valid?

Reply:
The `result` can be either Right if the validation passed, or Left if the validation fails. In this case, I called `PathReporter` to get the error message and `isRight` returns true or false depending on if the validation was successful or not.